runtime/critical_section.h

```diff
@@ -26,13 +26,17 @@ class CriticalSectionSmartGuard : private vk::not_copyable {
   CriticalSectionSmartGuard() noexcept { enter_critical_section(); }
 
   void leave_critical_section() noexcept {
-    dl::leave_critical_section();
-    in_critical_section_ = false;
+    if (in_critical_section_) {
+      dl::leave_critical_section();
+      in_critical_section_ = false;
+    }
   }
 
   void enter_critical_section() noexcept {
-    dl::enter_critical_section();
-    in_critical_section_ = true;
+    if (!in_critical_section_) {
+      dl::enter_critical_section();
+      in_critical_section_ = true;
+    }
   }
 
   ~CriticalSectionSmartGuard() noexcept {
```

Comment:
and now you can just call the `leave_critical_section` method here, without `if`